flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -652,20 +653,47 @@ private CompletableFuture<CleanupJobState> handleJobManagerRunnerResult(
                 && executionType == ExecutionType.RECOVERY) {
             return CompletableFuture.completedFuture(
                     jobManagerRunnerFailed(
-                            jobManagerRunnerResult.getExecutionGraphInfo().getJobId(),
+                            jobManagerRunnerResult,
                             jobManagerRunnerResult.getInitializationFailure()));
         }
         return jobReachedTerminalState(jobManagerRunnerResult.getExecutionGraphInfo());
     }
 
-    enum CleanupJobState {
-        LOCAL,
-        GLOBAL
+    private static class CleanupJobState {
+
+        private final JobStatus jobStatus;
+        private final boolean globalCleanup;
+
+        public static CleanupJobState localCleanup(JobStatus jobStatus) {
+            return new CleanupJobState(jobStatus, false);
+        }
+
+        public static CleanupJobState globalCleanup(JobStatus jobStatus) {
+            return new CleanupJobState(jobStatus, true);
+        }
+
+        private CleanupJobState(JobStatus jobStatus, boolean globalCleanup) {
+            this.jobStatus = jobStatus;
+            this.globalCleanup = globalCleanup;
+        }
+
+        public boolean isGlobalCleanup() {
+            return globalCleanup;
+        }
+
+        public JobStatus getJobStatus() {
+            return jobStatus;
+        }
     }
 
-    private CleanupJobState jobManagerRunnerFailed(JobID jobId, Throwable throwable) {
-        jobMasterFailed(jobId, throwable);
-        return CleanupJobState.LOCAL;
+    private CleanupJobState jobManagerRunnerFailed(
+            JobManagerRunnerResult jobManagerRunnerResult, Throwable throwable) {
```

Comment:
Small question: since you're pulling `getExecutionGraphInfo().getJobId()` and `.getExecutionGraphInfo().getArchivedExecutionGraph().getState()` from the result, would it make more sense to write the type signature as:

```
private CleanupJobState jobManagerRunnerFailed(
￼            ExecutionGraphInfo executionGraphInfo, Throwable throwable) {
```

Or is passing the result in better from a readability point of view?

Reply:
yikes, you have a point. But while looking into it I noticed that I did something stupid here: I used the `jobManagerRunnerResult` for `jobManagerRunnerFailed` in the case where `jobManagerRunnerResult` is actually `null`. I fixed this now. But I'm gonna double-check why this never came up in any test...

Reply:
I added a test case to cover this code path.